Supervision order REST endpoints

corrad82-4s opened the pull request:
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #2878
* Related to DSpace/RestContract#208 

## Description
A rest endpoint according to what defined in REST Contract, accepting CRUD operations for supervision orders, with usage limited to Administrators has been exposed. Two new discovery configurations have been added: "otherworkspace" returning all Workspace item a user is a member of their supervision group, and "supervision" returning all workspace and workflow items, used by UI Workflow admin page. 

## Instructions for Reviewers


List of changes in this PR:
Several files have been added and / or updated, please refer to list of Changed files for a complete list.

Our suggestion is to test this PR together with its [angular part](https://github.com/DSpace/dspace-angular/pull/2019) 



## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

Commit:
[CST-7754] LYRASIS: Supervisor orders (REST).

Commit:
[CST-7754] removed unused fields.

Commit:
[CST - 7754] refactoring and added support for NONE type

Commit:
Merge branch 'main' into CST-7754

Commit:
[CST-7754] otherworkspace discovery configuration shows only supervised workspace items, when a supervision order is deleted grants given to supervision groups on an item are revoked.

Commit:
[CST-7754] test supervision order is deleted if workspaceItem or group is deleted.

corrad82-4s commented:
We have updated this PR with an IT proving that when groups or items are deleted their related supervision orders are deleted as well

tdonohue commented on dspace/config/spring/api/discovery.xml:
```diff
@@ -894,6 +901,79 @@
     </bean>
 
+    <!--The other workspace configuration settings for discovery -->
+    <bean id="otherWorkspaceConfiguration"
```
What do we mean by an "otherWorkspaceConfiguration"?  Could we change the name here (the phrase "other workspace" is not at all clear)? How does this relate to or differ from the "supervisionConfiguration?

abollini replied:
maybe supervisedWorkspaceConfiguration as there is already a supervionConfiguration used by the administrative dashboard?

tdonohue reviewed (changes requested):
@corrad82-4s : I gave this a code review today, and I'm not sure I understand why this seems like a complete "redesign" of Supervision Orders from the way they worked in 6.x?

It appears in this PR that Supervision Orders are no longer limited to WorkspaceItems (i.e. the submission process) as they were in 6.x.  They now appear on WorkflowItems as well, and are linked directly to the **Item** object itself (and I'm not seeing anyplace they are ever removed once the Item goes to "in_archive" status... and no tests that prove the Supervision order goes away once the Item is archived).

It also appears that there is no migration process for migrating older Supervision Orders (from 6.x) to 7.x.  All the 6.x code appears to be replaced, but it hasn't been removed from the codebase, and the data is never migrated.

I think this entire design needs more discussion as it's not what I anticipated.  It's no longer simply recreating 6.x JPSUI Supervision Orders... Instead, it's a new type of Supervision Order, and I don't understand how it's expected to work (as there's no clear documentation that I can find) or the use cases behind it.

corrad82-4s commented:
Hi @tdonohue , 
thank you for your comments, we are reviewing what suggested and reported and will get back to you soon

corrad82-4s commented on dspace-api/src/main/java/org/dspace/supervision/SupervisionOrder.java:
```diff
@@ -0,0 +1,78 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "item_id")
+    private Item item;
```
This mainly was to be more adherent to what set in contract, where we expect receiving an uuid when a supervision order is set. IMHO referencing the Item can help us in logic where we are retrieving supervision orders or executing discovery queries to retrieve supervised workspace items. Of course we must be sure that supervision order, once the item is archived, is removed.

tdonohue replied:
@corrad82-4s ; In that case, I recommend adding a test to verify that the supervision order is *deleted* when the item changes state.  It seems like it should be deleted once the WorkspaceItem is removed (i.e. once the WorkspaceItem transitions into a WorkflowItem).  As it's unclear to me why the supervision order should even remain once the Item progresses into the Workflow approval state.  But, if there's a reason to still have the supervision order on the `WorkflowItem`, then we should **minimally** add a test to verify the Supervision order is deleted once the Item is archived.

corrad82-4s replied:
@tdonohue : Yes sorry, the supervision order must be deleted once WorkspaceItem transitions into a WorkflowItem, not when the Item is archived

abollini replied:
@corrad82-4s  @tdonohue  is right in saying that the supervisor order must be deleted once the item is archived but the supervisor order must be retained during the workflow for the reason explained here https://github.com/DSpace/DSpace/pull/8621/files#r1083021166 in my other reply

abollini commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceSupervisionOrderIndexingPlugin.java:
```diff
@@ -0,0 +1,64 @@
+        try {
+
+            if (!(indexableObject instanceof IndexableWorkspaceItem) &&
+                !(indexableObject instanceof IndexableWorkflowItem)) {
```
I have reviewed our initial notes about that. The reason why we need to keep track of the supervision also when the item is in workflow is that workflow item can be rejected back to the workspaceitem. The idea is that if an item was supervised it should be still supervised when it goes back. In version 6 this was managed in an inconsistent way as the resource policies related to the supervision was left in place and so available when the item was sent back but the workspaceitem2eperson link was of course loss.
Please note that keep the information about the supervision when the item is in workflow add no functional changes as the supervisors cannot do anything to the workflowitem. This can be eventually proven by a dedicated test

tdonohue replied:
@abollini : That makes sense as a use case for why a Supervision Order might remain on a `WorkflowItem`.  I can agree to that.

In that case, I'd recommend we create more tests which align with this use case / prove that it works as expected.  For instance, based on this use case, I'd assume the following should be true:
1. When an Item moves from WorkspaceItem to WorkflowItem, the SupervisionOrder is retained.  However, the users/groups related to the SupervisionOrder only have `READ` privileges on the Item when it is in Workflow (in the same way that a Submitter can only "read" their submission once it is in Workflow & can no longer modify it unless it is rejected & returned to "WorkspaceItem" status).  
    * In other words, while in WorkspaceItem status, a SupervisionOrder can allow for edit privileges. But, once in WorkflowItem status, it only allows for read privileges.
2. Obviously, when the Item becomes archived, the SupervisionOrder should be deleted.

tdonohue replied:
Closing this request... I've verified these new tests have been added.

abollini commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V7.5_2022.12.09__Supervision_Orders_table.sql:
```diff
@@ -0,0 +1,20 @@
+  id INTEGER PRIMARY KEY,
+  item_id UUID REFERENCES Item(uuid) ON DELETE CASCADE,
+  eperson_group_id UUID REFERENCES epersongroup(uuid) ON DELETE CASCADE
+);
```
The sql migration script is missing. In v6 we have the workspaceitem2eperson table that must be migrated to the new table. Moreover, checking the v6 code we have identified an issue, the resourcepolicies added for a supervisor order were never deleted, I suggest to create such policies with `TYPE_SUBMISSION` as by their nature are only relevant for the submission in the workspaceitem. In such way the policies will be automatically removed when the item is archived

abollini commented on dspace-api/src/main/java/org/dspace/app/exception/ResourceConflictException.java:
```diff
@@ -0,0 +1,39 @@
+ * @author Mohamed Eskander (mohamed.eskander at 4science dot it)
+ *
+ */
+public class ResourceConflictException extends RuntimeException {
```
not sure we really need that

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/exception/DSpaceApiExceptionControllerAdvice.java:
```diff
@@ -235,6 +248,12 @@ protected void handleGenericException(HttpServletRequest request, HttpServletRes
+    @ExceptionHandler(ResourceConflictException.class)
+    protected ResponseEntity<? extends RestModel> resourceConflictException(ResourceConflictException ex) {
+        RestModel resource = converterService.toRest(ex.getResource(), utils.obtainProjection());
+        return new ResponseEntity<RestModel>(resource, HttpStatus.CONFLICT);
```
I don't think we really need to introduce a new status here. This seems more related to 422 error code

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/model/WorkspaceItemRest.java:
```diff
@@ -14,10 +14,18 @@
+@LinksRest(links = {
+    @LinkRest(
+        name = WorkspaceItemRest.SUPERVISION_ORDERS,
+        method = "getSupervisionOrders"
```
see https://github.com/DSpace/DSpace/pull/8621/files#r1083021166

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SupervisionOrderRestRepository.java:
```diff
@@ -0,0 +1,239 @@
+
+        supervisionOrder = supervisionOrderService.findByItemAndGroup(context, item, group);
+        if (Objects.nonNull(supervisionOrder)) {
+            throw new ResourceConflictException(
```
by previous agreement this special kind of conflict related to business rule was decided to be managed with a 422 see for instance here https://github.com/DSpace/RestContract/pull/180#discussion_r850829376

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:
```diff
@@ -5987,4 +5991,491 @@ public void discoverFacetsTestWithDsoTypeTest() throws Exception {
+        SupervisionOrderBuilder.createSupervisionOrder(context, wsItem2Admin.getItem(), groupA).build();
+
+        SupervisionOrderBuilder.createSupervisionOrder(context, wfItem1Admin.getItem(), groupA).build();
+        SupervisionOrderBuilder.createSupervisionOrder(context, wfItem1Admin.getItem(), groupB).build();
```
I would like to support this use case at the least (initially) from the REST perspective.
An administrator can add a supervisor order to a workflowitem before reject it so that the supervisor can help in moving forward the submission

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/SupervisionOrderRestRepositoryIT.java:
```diff
@@ -0,0 +1,1088 @@
+                        .andExpect(jsonPath("$.inArchive", Matchers.is(false)));
+
+        getClient(adminToken).perform(get("/api/core/supervisionorders/" + supervisionOrder.getID()))
+                             .andExpect(status().isOk())
```
I agree with @tdonohue this should lead to an exception of illegalstate. On the rest side this should be a 422. The item to supervise must be an inprogress submission (workspace/workflow)

abollini reviewed (changes requested):
@corrad82-4s I have added my review to this PR as well. I take the chance to respond to some questions from @tdonohue about the scope/purpose of some decision, it looks to me that the code is almost ready but some cleanup and fixes are required

Commit:
[CST-7754] removed old code of supervision orders
added a new migration script sql
tested that supervisors can edit only workspace items

Commit:
Merge remote-tracking branch 'origin/main' into CST-7754

# Conflicts:
#	dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java

corrad82-4s commented:
@tdonohue , @abollini thank you for these reviews, they are all being addressed, updating the PR once ready

Commit:
[CST-7754] updated javadoc

Commit:
[CST-7754] updated comment

Commit:
[CST-7754] withdrawn item should not allow supervision order

Commit:
[CST-7754] moved supervision orders test on workflow items to SupervisionOrderRestRepositoryIT, added test covering the scenario when an admin tries to set a supervision order on a withdrawn item, and when an admin sets a supervision order on a workflow item before sending it back to workspace: supervisor can edit the item only when it goes to workspace status.

Commit:
Merge branch 'main' into CST-7754

corrad82-4s commented:
@tdonohue , @abollini we have addressed the feedback and updated the PR. It is ready for a review. In the next days. 
Please note that at least one of addressed feedback impacts on Angular PR, we will update it in the next days.

Commit:
[CST-7754] supervising user can perform operations on Workspace item only if they have proper permissions

tdonohue reviewed (commented):
@corrad82-4s : Just wanted to note that I've re-reviewed the code of this backend PR.  It all looks good to me now.  So, once the frontend bugs are resolved, I think the backend is ready to go.

Commit:
Merge branch 'main' into CST-7754

# Conflicts:
#	dspace-api/src/test/java/org/dspace/builder/AbstractBuilder.java

Commit:
Merge branch 'main' into CST-7754

# Conflicts:
#	dspace-api/src/main/java/org/dspace/content/InstallItemServiceImpl.java

Commit:
Merge branch 'main' into CST-7754

Commit:
Merge branch 'main' into CST-7754

# Conflicts:
#	dspace-api/src/test/java/org/dspace/builder/AbstractBuilder.java

abollini reviewed (approved):
thanks @corrad82-4s my previous feedback has been addressed, it looks good to me now

tdonohue reviewed (approved):
👍 Thanks @corrad82-4s !  My feedback has been addressed & I've verified this is working well with the updated frontend.
